packages/odyssey-react/src/components/TextInput/index.tsx

```diff
@@ -0,0 +1,159 @@
+ *
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *
+ * See the License for the specific language governing permissions and limitations under the License.
+ */
+
+import React, { useMemo } from 'react';
+import type { FunctionComponent, ChangeEvent, ReactNode } from 'react';
+import { nanoid } from 'nanoid';
+const buildId = nanoid.bind(null, 6);
+
+export type TextInputProps = {
+  /**
+   * The underlying input element id attribute. Automatically generated if not provided.
+   */
+  id?: string,
+
+  /**
+   * The underlying input element type.
+   * @default text
+   */
+  type?: 'text' | 'email' | 'url' | 'tel' | 'search' | 'password',
+
+  /**
+   * The form field label.
+   */
+  label: string,
+
+  /**
+   * The underlying input element required attribute.
+   * @default true
+   */
+  required?: boolean,
+
+  /**
+   * The underlying input element disabled attribute.
+   * @default false
+   */
+  disabled?: boolean,
+
+  /**
+   * The underlying input element readonly attribute.
+   * @default false
+   */
+  readonly?: boolean,
+
+  /**
+   * Text to display when the form is optional, i.e. required prop is false.
+   */
+  optionalLabel?: string,
+
+  /**
+   * The underlying input element placeholder attribute.
+   */
+  placeholder?: string,
+
+  /**
+   * The input element value for controlled components.
+   */
+  value?: string,
+
+  /**
+   * The initial input element value for uncontrolled components.
+   */
+  defaultValue?: string,
+
+  /**
+   * Callback executed when the input fires a change event.
+   * @param {Object} event the event object
+   * @param {string} value the string value of the input
+   */
+  onChange?: (event: ChangeEvent<HTMLInputElement>, value: string) => void,
+
+  // onBlur: function,
+
+  // onFocus: function,
+};
+
+
+function renderLabel(
+  label: string,
+  required: boolean,
+  optionalLabel?: string
+): ReactNode {
+  if (required || !optionalLabel) {
+    return <>{label}</>;
+  }
+  return (
+    <>
+      {label}
+      <span
+        className="ods-label--optional"
+        children={optionalLabel}
+      />
+    </>
+  );
+}
+
+function buildHandleChange(onChange: TextInputProps['onChange']) {
+  return function (event: ChangeEvent<HTMLInputElement>) {
+    onChange?.(event, event.target.value);
+  };
+}
+
+/**
+ * Text inputs allow users to edit and input data.
+ */
+const TextInput: FunctionComponent<TextInputProps> = (props) => {
+  const {
+    defaultValue,
+    disabled = false,
+    id = useMemo(buildId, []),
+    label,
+    onChange,
+    optionalLabel,
+    placeholder,
+    readonly = false,
+    required = true,
+    type = 'text',
+    value,
```

Comment:
While they're present in the doc examples, Hints and Errors are currently part of a larger Field component. I'm 💯 on leaving them out here, since we'll need them to be separate components used across other field types as well (e.g. Radio).

Reply:
Thanks - yes we need to lock in the API for how these components behave. I have some ideas to put forward to the group around this topic that should be flexible enough for broad application across components. Perhaps we discuss at our next craft session?

I would like the API of these components to be mostly finalized out as part of our MVP effort over the next two months.